File: Day25.hs
import Control.Monad (forM_)
import Data.List (elemIndex,unfoldr)
import Data.Maybe (fromJust)

day25 :: [String] -> String
day25 = reverse . unfoldr op . sum . map decypher
  where digits = "=-012"
        decypher = foldl (\res c -> res * 5 + (fromJust $ elemIndex c digits) - 2) 0
        op 0 = Nothing
        op n = Just (digits !! (m + 2), (n - m) `div` 5)
          where m = ((n `mod` 5) + 2) `mod` 5 - 2 -- map [0..4] to [0,1,2,-2,-1]

main :: IO ()
main = forM_ ["input/25test.txt", "input/25full.txt"] $ \filename -> do
        input <- lines <$> readFile filename
        print $ day25 input

File: Day25.cpp
#include "Lines.h"
#include <fmt/core.h>
#include <array>
#include <algorithm> // std::ranges::find

const std::array<const char, 5> digits = { '=','-','0','1','2' };

int64_t decypher(std::string_view str) {
	int64_t res = 0;
	for (const char c : str) { // This is std::ranges::fold
		res = res * 5 + (std::ranges::find(digits, c) - digits.begin() - 2);
	}
	return res;
}

void day25(const char* filename) {
	int64_t sum = ranges::sum(lines(filename), decypher);
	std::string res = {};
	while (sum) { // This is Data.List.unfoldr :)
		const int64_t mod = ((sum % 5) + 2) % 5 - 2;
		res.push_back(digits[mod + 2]);
		sum -= mod;
		assert(sum % 5 == 0);
		sum /= 5;
	}
	std::ranges::reverse(res);
	fmt::print("{}\n", res);
}

int main() {
	day25("input/25test.txt"); // 2=-1=0
	day25("input/25full.txt"); // 2-121-=10=200==2==21
	return 0;
}
